src/Features/Core/Portable/EmbeddedLanguages/Json/JsonNodes.cs

```diff
@@ -0,0 +1,311 @@
+
+        internal override int ChildCount => 2;
+
+        internal override JsonNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => MinusToken,
+                1 => LiteralToken,
+                _ => throw new InvalidOperationException(),
+            };
+
+        public override void Accept(IJsonNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    internal sealed class JsonLiteralNode : JsonValueNode
+    {
+        public JsonLiteralNode(JsonToken literalToken)
+            : base(JsonKind.Literal)
+        {
+            LiteralToken = literalToken;
+        }
+
+        public JsonToken LiteralToken { get; }
+
+        internal override int ChildCount => 1;
+
+        internal override JsonNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => LiteralToken,
+                _ => throw new InvalidOperationException(),
+            };
+
+        public override void Accept(IJsonNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    internal sealed class JsonCommaValueNode : JsonValueNode
+    {
+        public JsonCommaValueNode(JsonToken commaToken)
+            : base(JsonKind.CommaValue)
+        {
+            Debug.Assert(commaToken.Kind == JsonKind.CommaToken);
+            CommaToken = commaToken;
+        }
+
+        public JsonToken CommaToken { get; }
+
+        internal override int ChildCount => 1;
+
+        internal override JsonNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => CommaToken,
+                _ => throw new InvalidOperationException(),
+            };
+
+        public override void Accept(IJsonNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    internal sealed class JsonPropertyNode : JsonValueNode
+    {
+        public JsonPropertyNode(JsonToken nameToken, JsonToken colonToken, JsonValueNode value)
+            : base(JsonKind.Property)
+        {
+            Debug.Assert(nameToken.Kind == JsonKind.StringToken || nameToken.Kind == JsonKind.TextToken);
+            Debug.Assert(colonToken.Kind == JsonKind.ColonToken);
+            Debug.Assert(value != null);
+            NameToken = nameToken;
+            ColonToken = colonToken;
+            Value = value;
+        }
+
+        public JsonToken NameToken { get; }
+        public JsonToken ColonToken { get; }
+        public JsonValueNode Value { get; }
+
+        internal override int ChildCount => 3;
+
+        internal override JsonNodeOrToken ChildAt(int index)
+            => index switch
+            {
+                0 => NameToken,
+                1 => ColonToken,
+                2 => Value,
+                _ => throw new InvalidOperationException(),
+            };
+
+        public override void Accept(IJsonNodeVisitor visitor)
+            => visitor.Visit(this);
+    }
+
+    internal sealed class JsonConstructorNode : JsonValueNode
```

Comment:
I really have no idea what a Json constructor is. Is it common? Docs would be helpful :)

Reply:
i'll doc.  it's a json.net extension to json :)

Reply:
done.